### PR TITLE
Add `--depth` option

### PR DESCRIPTION
**Pull request by BennD**
added `--depth` option
removed `--direct-dependencies-only` option, as it is equivalent to `--depth 1`

### CODE REVIEW

File: src/graph/build.rs [hunk 1]
```diff
@@ -99,9 +99,9 @@ pub(crate) fn get_dep_graph(metadata: Metadata, config: &Config) -> anyhow::Resu
 
                     let idx = graph.add_node(dep_pkg);
 
-                    // Don't add any more dependencies to the queue if we only want direct dependencies
-                    if !config.direct_dependencies_only {
-                        deps_add_queue.push_back(dep.pkg.clone());
+                    // Don't add dependencies of dependencies if we're at the depth limit
+                    if depth + 1 < config.depth.unwrap_or(u32::MAX) {
+                        deps_add_queue.push_back((dep.pkg.clone(), depth + 1));
                     }
```
Comment (jplatte) [inline src/graph/build.rs:105]: Thinking more about this, workspace-only seems like a special case of `depth`. I don't think it should be removed here in favor of `--depth=0`, but I wonder why workspace-only is handled differently than this. If you don't have an answer that's fine, I guess that means I should just revisit this code after merging the PR but before releasing.

Comment (BennD) [reply src/graph/build.rs:105]: While thinking about this i noticed that `--depth 0` is actually broken.
I will take a look at this either tomorrow or the day after.

Comment (jplatte) [reply src/graph/build.rs:105]: Great, thanks :)

**Comment by BennD** [comment]
My new version of `--depth` works with a depth of `0` now and also works more like `--workspace-only`.
I removed the `is_workspace_member` check, as only non-workspace crates end if the
`HashMapEntry::Vacant(v) => {...}` pattern. Workspace crates are all already added in line 48.

To simplify the code and reduce redundancy, we could remove the `--workspace_only` check and
instead set `--depth` to `0` if the `--workspace-only` option is used.
In case `--workspace-only` and `--depth` are defined, `--workspace-only` should override whatever is set with `--depth`.
Maybe issue a warning though? Or just fail and tell the user that those options are not compatible?

**Comment by jplatte** [comment]
Re. removal of `is_workspace_member`, are you sure about that? I think if the user explicitly sets a root package, workspace members *aren't* all added at the same time.

**Comment by BennD** [comment]
> Re. removal of `is_workspace_member`, are you sure about that? I think if the user explicitly sets a root package, workspace members _aren't_ all added at the same time.

You are correct. That also means `--workspace-only` is not equal to `--depth 0` as initially thought.

**Comment by jplatte** [comment]
Oh, as in `--workspace-only` combined with `--root foo` is different? That is true, maybe that's a reason to keep both flags, not just for now but indefinitely.

File: src/graph/build.rs [hunk 2]
```diff
@@ -88,6 +88,11 @@ pub(crate) fn get_dep_graph(metadata: Metadata, config: &Config) -> anyhow::Resu
                         continue;
                     }
 
+                    // Don't add dependencies of dependencies if we're at the depth limit
+                    if depth + 1 > config.depth.unwrap_or(u32::MAX) {
```
Comment (jplatte) [inline src/graph/build.rs:92]: This should work too, right? Seems a bit more intuitive to me.

```suggestion
                    if config.depth.is_some_and(|max_depth| depth >= max_depth) {
```

Comment (BennD) [reply src/graph/build.rs:92]: Yes thats definitely a better way to do it.

**Review by jplatte — APPROVED** [review]
Thanks!

**Pull request merged**